mark pointer inputs with flag

DieracDelta opened the pull request:
This PR introduces a flag, `--wp-pointer-reg-list` and closes the second half of #216 . The purpose of this flag is to, at the start of a function, mark registers as pointers. These specially marked registers are then used to generate a precondition that restricts their value to anywhere that is not uninitialized stack. That is to say, these "pointer" register values are either above the initial value of `RSP` or below the bottom of the stack. A sample invocation is included with the example, [here](https://github.com/DieracDelta/cbat_tools/blob/jrestivo/precond_flag/wp/resources/sample_binaries/pointer_flag/run_wp_unsat.sh).

I've tested this on the grammatech examples from #213 , and the simple toy example included in this commit.
- The toy example: CBAT is unable to use differing return addresses to obtain differing return values with the flag (the inspiration for this PR), but comes back SAT without the flag.
- `is_dumpdir` remains SAT when `RDI` is marked as a pointer. This is because in execution, there are accesses at an offset, which result in `wp` choosing a particular `RSP` such that `RDI + $OFFSET == pushed return address`.
- `safe_hasher` and `mb_copy` suffer from the same issue as `is_dumpdir`, though they're more subtle.

Note that I've only included a test for the comparative case (and double checked that the single analysis case doesn't break things). It's not obvious to me how to obtain different results (SAT vs UNSAT) with and without this flag in a single binary analysis.

DieracDelta commented:
I've taken @codyroux 's suggestions into account. Specific changes include:
- A rebase onto master (using the new cli)
- A single binary analysis test case for the flag
- Constraints are now constructed using the Z3 api only
- Comments are included in the `construct_pointer_constraint` docstring detailing the constraint being built

fortunac commented on wp/lib/bap_wp/src/z3_utils.ml:
```diff
@@ -79,13 +79,13 @@ let build_str (tokens : string list) : string =
    init_RDI_orig. *)
 let get_z3_name (map : Constr.z3_expr Var.Map.t) (name : string) (fmt : Var.t -> string)
-  : string option =
+  : (Z3.Expr.expr option) =
```
Super small nit, can we use `Constr.z3_expr` here to be more consistent?

fortunac commented on wp/lib/bap_wp/src/z3_utils.mli:
```diff
@@ -32,7 +32,7 @@ val build_str : string list -> string
     init_RDI_orig. *)
 val get_z3_name :
-  Constr.z3_expr Bap.Std.Var.Map.t -> string -> (Bap.Std.Var.t -> string) -> string option
+  Constr.z3_expr Bap.Std.Var.Map.t -> string -> (Bap.Std.Var.t -> string) -> (Z3.Expr.expr option)
```
Here too.

fortunac commented on wp/plugin/lib/parameters.ml:
```diff
@@ -35,6 +36,8 @@ type t = {
   show : string list
 }
 
+(* TODO add in validate pointer_reg_list *)
```
Are you planning on adding this function in this PR or later?

DieracDelta replied:
Oops! Since this works on both single and comparative analysis, it's probably not necessary? I deleted the comment.

fortunac commented on wp/resources/sample_binaries/pointer_flag/run_wp_unsat.sh:
```diff
@@ -0,0 +1,24 @@
+# is thereby pushed onto the stack to cause a different return value between functions.
+# Should return SAT
+
+# Should return UNSAT
```
Are we expecting a SAT or UNSAT for this test?

fortunac commented on wp/lib/bap_wp/src/z3_utils.ml:
```diff
@@ -111,3 +111,58 @@ let mk_smtlib2_single (env : Env.t) (smt_post : string) : Constr.t =
+  let arch = match Env.get_arch env1 |> Bap.Std.Arch.addr_size with
+    | `r32 -> 32
+    | `r64 -> 64 in
+  let rsp = "RSP" in
```
Hardcoding the string "RSP" makes me a bit uncomfortable. Since we have the architecture, we can grab the name of the stack pointer:
```
let module Target = (val target_of_arch arch) in
let rsp = Var.name Target.CPU.sp in
...
```

fortunac commented on wp/lib/bap_wp/src/environment.mli:
```diff
@@ -251,6 +251,9 @@ val use_input_regs : t -> bool
     defined by the concrete range of the stack in the env. *)
 val in_stack : t -> Constr.z3_expr -> Constr.z3_expr
 
+(** [get_stack_bottom] retrieves the address of the bottom of the stack *)
```
Can you add a bit of clarification that the stack is grows from top to bottom, and the stack bottom is not the same as its base?

DieracDelta commented:
I've incorporated in @fortunac 's suggestions. A couple things I'm uncertain about:
- the explanation for `get_stack_bottom` is a bit confusing. I renamed the function to `get_stack_end` and changed comments. Hopefully this is less confusing, thought it's still a bit of a misnomer, since it's not exactly the end of the stack as much as the minimum value that the stack pointer can take/the largest size the stack can be.
- The placement of `construct_pointer_constraint` is in `z3_utils`. It's making ample use of the Z3 api, which is why I've placed it here.

fortunac reviewed (commented):
Looks good to me! Just a few minor details that I commented below. Cody might want to double check the z3 expression generated, but it looks good to me. Good job!

DieracDelta commented:
Added docs and fixed those makefiles. Good catch!

codyroux commented on wp/lib/bap_wp/src/compare.mli:
```diff
@@ -104,6 +104,20 @@ val compare_subs_empty_post : comparator * comparator
     creation time. *)
 val compare_subs_sp : comparator * comparator
 
+(** Compare two subs by composition for restraints on a pointer.:
```
I don't think this is an appropriate description of this function: nothing in the body of the function mentions pointers in any way.

In general this function seems very generic, so I'm not sure why you're mentioning x86_64, or indeed why the description is what it is. It looks like this function just bottles up pre and post conditions. The comment should reflect that.

codyroux commented on wp/lib/bap_wp/src/z3_utils.ml:
```diff
@@ -111,3 +111,59 @@ let mk_smtlib2_single (env : Env.t) (smt_post : string) : Constr.t =
+    | `r32 -> 32
+    | `r64 -> 64 in
+  let rsp = Env.get_rsp_name env1 in
+  let err_msg_input = "invalid register name in pointer register list" in
```
Probably a good idea to have the name of the function in the error message, in my experience.

codyroux replied:
Like `"construct_pointer_constraint: invalid register name in pointer register list"`

codyroux commented on wp/lib/bap_wp/src/z3_utils.ml:
```diff
@@ -111,3 +111,59 @@ let mk_smtlib2_single (env : Env.t) (smt_post : string) : Constr.t =
+    | `r64 -> 64 in
+  let rsp = Env.get_rsp_name env1 in
+  let err_msg_input = "invalid register name in pointer register list" in
+  let err_msg_rsp = "stack pointer not found" in
```
And same here.

codyroux commented on wp/lib/bap_wp/src/z3_utils.ml:
```diff
@@ -111,3 +111,59 @@ let mk_smtlib2_single (env : Env.t) (smt_post : string) : Constr.t =
+             (get_z3_name init_var_map_orig reg Var.name) in
+         let reg_name_mod = Option.value_exn ~message:err_msg_input
+             (get_z3_name init_var_map_mod reg Var.name) in
+         (* the pointer register must be above RSP
```
As a side note, if you can get your hands on a module of type `CPU`, you can use this function: https://binaryanalysisplatform.github.io/bap/api/odoc/bap/Bap/Std/module-type-CPU/index.html#val-sp to get your hands on the actual variable that contains the top of the stack.

DieracDelta replied:
Yes! I added this into `environment.ml` as `get_rsp_name`. I realize the name is x86 specific, so I'm going to change the name.

codyroux commented on wp/lib/bap_wp/src/z3_utils.ml:
```diff
@@ -111,3 +111,59 @@ let mk_smtlib2_single (env : Env.t) (smt_post : string) : Constr.t =
+         (* the pointer must be below the end of the stack *)
+         let ule_1 =  Z3.BitVector.mk_ule ctx reg_name_orig sb_bv in
+         let ule_2 =  Z3.BitVector.mk_ule ctx reg_name_mod sb_bv in
+         (* encode that the pointer is either above RSP or below the stack
```
Great! You might want to write this down as a formula, like `RSP <= R \/ R >= stack_end` or something. I find formulae to be slightly more readable in this context.

No big deal though.

codyroux commented on wp/lib/bap_wp/src/z3_utils.ml:
```diff
@@ -111,3 +111,59 @@ let mk_smtlib2_single (env : Env.t) (smt_post : string) : Constr.t =
+         let and_c = Z3.Boolean.mk_and ctx [or_c_1; or_c_2;] in
+         Z3.Boolean.mk_and ctx [and_c; acc]
+      )
+    | None ->
```
Maybe worth commenting that this is the case where we're analyzing a single binary.

codyroux commented on wp/plugin/lib/analysis.ml:
```diff
@@ -138,6 +138,19 @@ let sp (arch : Arch.t) : (Comp.comparator * Comp.comparator) option =
            memory for the x86_64 architecture.\n%!";
     None
 
+
```
Wait, why do we have a `lib` sub-folder in our `plugin` folder? Shouldn't this be in `wp/lib`?

codyroux replied:
Nevermind.

codyroux commented on wp/plugin/lib/analysis.ml:
```diff
@@ -151,7 +164,8 @@ let comparators_of_flags
       ~orig:(sub1, env1) ~modif:(sub2, env2);
     smtlib ~precond:p.precond ~postcond:p.postcond;
-    sp arch
+    sp arch;
```
Hey there's the stack pointer!

codyroux commented on wp/resources/sample_binaries/pointer_flag/main_1.c:
```diff
@@ -0,0 +1,15 @@
+int f(){
+    return 1;
+}
+// Type your code here, or load an example.
```
What does this comment mean?

codyroux commented on wp/resources/sample_binaries/pointer_flag/run_wp_sat.sh:
```diff
@@ -0,0 +1,23 @@
+# This is a test that tests that the wp-pointer-reg-list flag works as expected.
+# Specifically, the function foo tries to use differing
```
I see no function called `foo`...

codyroux commented on wp/resources/sample_binaries/pointer_flag/run_wp_sat.sh:
```diff
@@ -0,0 +1,23 @@
+# This is a test that tests that the wp-pointer-reg-list flag works as expected.
+# Specifically, the function foo tries to use differing
+# return addresses that are pushed onto the stack to cause a different return
+# value between functions. The abscence of the flag should allow the pointer
```
It's spelled "absence". Not to worry, my spelling is atrocious, but you might want to run some quick spell checks, or enable the squigglies in your IDE.

codyroux commented on wp/resources/sample_binaries/pointer_flag/run_wp_unsat.sh:
```diff
@@ -0,0 +1,24 @@
+# This is a test that tests that the wp-pointer-reg-list flag works as expected.
```
A test that tests the test.

codyroux commented on wp/resources/sample_binaries/pointer_flag/run_wp_sat.sh:
```diff
@@ -0,0 +1,23 @@
+# argument in RDI to be used to point to this return address, as it lives
+# on the uninitialized portion of the stack at the start of the function.
+
+# Should return SAT
```
I actually don't get how this example is supposed to work. What's happening with `num1`? It might help to actually refer to the C names of variables and functions.

codyroux commented on wp/resources/sample_binaries/pointer_flag/main_1.c:
```diff
@@ -0,0 +1,15 @@
+}
+// Type your code here, or load an example.
+int deref(int *num1) {
+    int num2 = f();
```
you could have `int num2 = 12;` and then add a `int local = 42;` right after, that way you need to prove that `*num1` doesn't point to the local stack.

This is the main problem we're having anyway.

DieracDelta commented:
I've made changes to incorporate in @codyroux 's suggestions. Specifically:
- Simplified the example in `pointer_flag`
- Various comment corrections
- Better error messages
- Fix a slight typo in the makefile

codyroux commented:
All clear from me!
